source/common/http/http2/codec_impl.cc

```diff
@@ -130,11 +130,14 @@ template <typename T> static T* removeConst(const void* object) {
 }
 
 ConnectionImpl::StreamImpl::StreamImpl(ConnectionImpl& parent, uint32_t buffer_limit)
-    : parent_(parent), local_end_stream_sent_(false), remote_end_stream_(false),
-      data_deferred_(false), received_noninformational_headers_(false),
+    : parent_(parent), buffer_memory_account_(std::make_shared<BufferMemoryAccount>()),
```

Comment:
I think you're creating BufferMemoryAccount for both server and client streams.  I think that the goal is to create BufferMemoryAccount on downstream, server streams only.  The upstream client streams should charge against the account owned by the downstream stream object.